Framework/Kernel/inc/MantidKernel/Cache.h

```diff
@@ -123,6 +167,7 @@ template <class KEYTYPE, class VALUETYPE> class DLLExport Cache {
     bool isValid = it_found != m_cacheMap.end();
 
     if (isValid) {
+      std::lock_guard<std::mutex> lock(m_mutex);
```

Comment:
After looking at the documentation you linked I think this might be fine even without the mutex? Insertion will not invalidate iterators (if I understood correctly), so this should be fine?